forge-game/src/main/java/forge/game/card/CardZoneTable.java

```diff
@@ -81,7 +86,8 @@ protected Table<ZoneType, ZoneType, CardCollection> delegate() {
 
     public void triggerChangesZoneAll(final Game game, final SpellAbility cause) {
         triggerTokenCreatedOnce(game);
-        if (!isEmpty()) {
+        if (!isEmpty() &&
```

Comment:
I would probably rewrite it to pass the ChangesZone Table as Replacement Parameter

With this, the RE doesn't need to create their own Table,  but can fill the existing one

Reply:
the problem I see there:
1. if the inner RE effect reuses the "parent" table we'd need a different sort of check in all outer places so it doesn't create a second but wrong entry?
2. and the other way around (maybe rarer) - if the RE gets replaced by another RE we'd also need to make sure only the zone changer at the bottom of that chain actually ends up in the table?

Reply:
you pass the table via params to the deepest GameAction moveZone part

then only the final move should be put into the table

Reply:
- `CardZoneTable.put()` will still get called by the outer "cause" after the RE is done
- so it should bail if entry origin/*/card already exists for that to work?
- then we need to be sure no weird case exists where a Table could **legally** have two such entries like some simultaneous stuff - Imo it's probably safe but what you think? 🤔 

I'd say the current solution is a bit cleaner code wise but `getCardState()` isn't that great so passing the table might be worth it.

Reply:
> * `CardZoneTable.put()` will still get called by the outer "cause" after the RE is done
and this is what i mean should be moved into GameAction

Currently:

Effect: 
* create Table
* getOldZone
* move Card
* compare OldZone with NewZone
* add to Table
* Trigger Table

what i want:

Effect:
* create Table (if Effect is not RE, we have similar checks for Cause)
* move Card (using Table as mapParam)
* Trigger Table (if Effect is not RE)

Game Action:
* calls ReplacementEffect
* if nothing replaces it, add Card to ZoneTable via the Param **This is the Part i want to add**

Reply:
Ok, I think I follow
will try later

Reply:
I updated ChangeZoneEffect as PoC